test/integration/cds_integration_test.cc

```diff
@@ -222,6 +222,117 @@ TEST_P(CdsIntegrationTest, CdsClusterTeardownWhileConnecting) {
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized"), nullptr);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+
+  // No cluster_2 stats yet.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_2.ClusterTrafficStats.initialized")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_2.upstream_cx_total"), nullptr);
+}
+
+// Test that LazyInitTrafficStats with cluster_1 create-remove-create sequence.
+TEST_P(LazyInitClusterStatsTest, LazyInitTrafficStatsWithClusterCreateDeleteRecrete) {
+  initialize();
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+
+  sendRequestToCluster1AndWaitForResponse();
+  // Cluster_1 trafficStats updated.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+  // Remove "cluster_1".
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster, {}, {},
+                                                             {ClusterName1}, "49");
+  test_server_->waitForCounterGe("cluster_manager.cluster_removed", 1);
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 2);
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                             {cluster2_}, {cluster2_}, {}, "43");
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 3);
+  EXPECT_EQ(test_server_->counter("cluster_manager.cluster_added")->value(), 3);
+  // No cluster_2 stats yet.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_2.ClusterTrafficStats.initialized")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_2.upstream_cx_total"), nullptr);
+  // Now the cluster_1 stats are gone, as well as the lazy init wrapper.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized"), nullptr);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+  // Now add cluster1 back.
+  updateCluster1();
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 4);
+  // Now the cluster_1.ClusterTrafficStats.initialized gauge is 0, since it didn't see previous
+  // stats.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+
+  // cluster_1 traffic stats created, due to the above http request.
+  sendRequestToCluster1AndWaitForResponse();
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+}
+
```

Comment:
so here's where I think the factoring out of stanzas is needed. I think you can probably make these tests about half the lines the are with a few easy refactorings, such as where you remove all clusters and wait for that to take effect.

Then I think you should repeat the tests you have, but with lazy-init off. I like repeating it better than using a parameter as I think you are really testing for different outcomes. The earlier versions of this where you parameterized whether to lazy-init were harder to follow IMO because of all the conditionals.

Reply:
shall we consider not testing the "non-lazyinit" case? I think those are well tested, and let this LazyInitClusterStatsTest test only focus on lazyinit stats related test?

Reply:
I'd prefer to test it, because we can specifically look for non-lazy behavior on the same stats.

Reply:
addded two more non-lazy test by copying the two lazy tests and removing non-related parts.